include/nbl/asset/ICPUBuffer.h

```diff
@@ -6,73 +6,91 @@
+    void* data = nullptr;
+    size_t alignment = _NBL_SIMD_ALIGNMENT;
+    std::pmr::memory_resource* memoryResource = nullptr;
+};
+
 //! One of CPU class-object representing an Asset
+//! TODO: remove, alloc can fail, should be a static create method instead!
 /**
     One of Assets used for storage of large arrays, so that storage can be decoupled
     from other objects such as meshbuffers, images, animations and shader source/bytecode.
 
     @see IAsset
 */
-class ICPUBuffer : public asset::IBuffer, public IPreHashed
+class ICPUBuffer final : public asset::IBuffer, public IPreHashed
 {
-    protected:
-        //! Non-allocating constructor for CCustormAllocatorCPUBuffer derivative
-        ICPUBuffer(size_t sizeInBytes, void* dat) : asset::IBuffer({ dat ? sizeInBytes : 0,EUF_TRANSFER_DST_BIT }), data(dat) {}
-
     public:
-        //! Constructor. TODO: remove, alloc can fail, should be a static create method instead!
-        /** @param sizeInBytes Size in bytes. If `dat` argument is present, it denotes size of data pointed by `dat`, otherwise - size of data to be allocated.
-        */
-        ICPUBuffer(size_t sizeInBytes) : asset::IBuffer({0,EUF_TRANSFER_DST_BIT})
-        {
-            data = _NBL_ALIGNED_MALLOC(sizeInBytes,_NBL_SIMD_ALIGNMENT);
-            if (!data) // FIXME: cannot fail like that, need factory `create` methods
-                return;
+        ICPUBuffer(size_t size, void* data, std::pmr::memory_resource* memoryResource, size_t alignment, bool adopt_memory) :
+            asset::IBuffer({ size, EUF_TRANSFER_DST_BIT }), m_data(data), m_mem_resource(memoryResource), m_alignment(alignment), m_adopt_memory(adopt_memory) {}
 
-            m_creationParams.size = sizeInBytes;
+        //! allocates uninitialized memory, copies `data` into allocation if `!data` not nullptr
+        core::smart_refctd_ptr<ICPUBuffer> static create(const SICPUBufferCreationParams& params) {
+            std::pmr::memory_resource* memoryResource = params.memoryResource;
+            if (!params.memoryResource)
+                memoryResource = std::pmr::get_default_resource();
+
+            auto data = memoryResource->allocate(params.size, params.alignment);
+            if (!data)
+                return nullptr;
+
+            if (params.data)
+                memcpy(data, params.data, params.size);
+
+            return core::make_smart_refctd_ptr<ICPUBuffer>(params.size, data, memoryResource, params.alignment, false);
+        }
+
+        //! does not allocate memory, adopts the `data` pointer, no copies done
+        core::smart_refctd_ptr<ICPUBuffer> static create(const SICPUBufferCreationParams& params, core::adopt_memory_t) {
+            std::pmr::memory_resource* memoryResource;
+            if (!params.memoryResource)
+                memoryResource = std::pmr::get_default_resource();
+            return core::make_smart_refctd_ptr<ICPUBuffer>(params.size, params.data, memoryResource, params.alignment, true);
         }

```

Comment:
take the struct as an r-value reference, then you can overwrite/modify its members (e.g. replace memory resource by the default one)

Reply:
also because of the requirement that the `memoryResource` be a smart-pointer, you must make global functions
```cpp
NBL_API2 nbl::core::refctd_memory_resource> getNullMemoryResource();
NBL_API2 nbl::core::refctd_memory_resource* getDefaultMemoryResource();
NBL_API2 void setDefaultMemoryResource(nbl::core::refctd_memory_resource*);
```

the implementations must use a `static smart_refctd_ptr<refctd_memory_resource>` inside the function  and re-initialize (`new` the default resource) it if its `nullptr`

Also to not annoy ASAN, we  may have to declare a DLL deinitialization section to DELETE those properly